### ISSUE TITLE
blendMode switching in FF

### ISSUE THREAD

**Issue by epar** [processing/p5.js#1756]
Switching between blendMode(ADD) and blendMode(BLEND) doesn't seem to work correctly in Firefox, both desktop and mobile. Once blendMode(ADD) is called it stays in ADD and doesn't go back to BLEND. 

All other browsers I've tried work fine.

I'm using p5 0.5.6

**Comment by epar** [comment]
Try the following. Works correctly on Chrome, Opera, Safari, but not on FF, all on OS X btw.

```
var x=0;

function setup() {
  createCanvas(600, 350);
  rectMode(CENTER);
  strokeWeight(5);
  stroke(255, 120);
  noFill();
}

function draw() {
  blendMode(BLEND);
  background(0);

  line(x, 50, x+50, 100);
  line(x, 100, x+50, 50);

  blendMode(ADD);
  line(x, 150, x+50, 200);
  line(x, 200, x+50, 150);

  blendMode(BLEND);
  line(x, 250, x+50, 300);
  line(x, 300, x+50, 250);

  x++;
  if(x>width) x=0;
}
```

**Comment by tstefanich** [comment]
I need to look deeper into it, but below is a quick bandaid that should get it to work in FF. If you reset blendmode(REPLACE) between switching to another blendmode() you should get it to render correctly. I'll dig deeper and try to have a proper fix.  

`var x=0;

function setup() {
  createCanvas(600, 350);
  rectMode(CENTER);
  strokeWeight(5);
  stroke(255, 120);
  noFill();
}

function draw() {
  background(0,0,0);
  blendMode(BLEND);
  line(x, 50, x+50, 100);
  line(x, 100, x+50, 50);
	blendMode(REPLACE);
  
	blendMode(ADD);
  line(x, 150, x+50, 200);
  line(x, 200, x+50, 150);
	blendMode(REPLACE);

	
  blendMode(BLEND);
  line(x, 250, x+50, 300);
  line(x, 300, x+50, 250);
	blendMode(REPLACE);

  x++;
  if(x>width) x=0;
}`

**Comment by nirbenyair** [comment]
We'll try to tackle it.

**Comment by NirBenya** [comment]
We'll tackle it.

**Comment by NirBenya** [comment]
We've succeed in replicating this outside of p5.js, this code works in Chrome but doesn't in FF:

```js
<script>
      window.requestAnimFrame = (function(callback) {
        return window.requestAnimationFrame || window.webkitRequestAnimationFrame || window.mozRequestAnimationFrame || window.oRequestAnimationFrame || window.msRequestAnimationFrame ||
            function(callback) {
              window.setTimeout(callback, 1000 / 60);
            };
      })();

      function drawRectangle(myRectangle, context) {
        context.beginPath();
        context.rect(myRectangle.x, myRectangle.y, myRectangle.width, myRectangle.height);
        context.fillStyle = 'blue';
        context.fill();
        context.lineWidth = myRectangle.borderWidth;
        context.strokeStyle = 'black';
        context.stroke();

      }
      function animate(myRectangle, canvas, context, startTime){
        // update
        var time = (new Date()).getTime() - startTime;

        var linearSpeed = 100;
        // pixels / second
        var newX = linearSpeed * time / 1000;

        if(newX < canvas.width - myRectangle.width - myRectangle.borderWidth / 2) {
          myRectangle.x = newX;
          myRectangle2.x = newX;
          myRectangle3.x = newX;

        }

        // clear
        context.fillStyle = 'black';
        context.fillRect(0, 0, canvas.width, canvas.height);
        context.globalCompositeOperation = "normal";

        drawRectangle(myRectangle, context);

        context.globalCompositeOperation = "lighter";

        drawRectangle(myRectangle2, context);

        context.globalCompositeOperation = "normal";

        drawRectangle(myRectangle3, context);

        // request new frame
        requestAnimFrame(function() {
          animate(myRectangle, canvas, context, startTime);

        });

      }
      var canvas = document.getElementById('myCanvas');
      var context = canvas.getContext('2d');
      canvas.style.background = 'black';
      canvas.width = '1000';
      canvas.height = '1000';
      var myRectangle = {
        x: 0,
        y: 75,
        width: 100,
        height: 50,
        borderWidth: 5
      };
      var myRectangle2 = {
        x: 0,
        y: 90,
        width: 100,
        height: 50,
        borderWidth: 5
      };
      var myRectangle3 = {
        x: 0,
        y: 300,
        width: 100,
        height: 50,
        borderWidth: 5
      };



      drawRectangle(myRectangle, context);


      // wait one second before starting animation
      setTimeout(function() {
        var startTime = (new Date()).getTime();
        animate(myRectangle, canvas, context, startTime);
      }, 1000);
    </script>
```

**Comment by lmccart** [comment]
aha, so should we file a bug report with FF?

**Comment by NirBenya** [comment]
The problem is that `normal` is not supported as a `globalCompositeOperation` in FF.
The corresponding operation is `source-over`. I'm not sure there is a difference between the two, but there may be when an image has an alpha value.

What could be done is to try and change it to normal, reread the value (which doesn't change when it's FF), and if it didn't change, change it to source-over.

#GoodnessSquad

**Comment by NirBenya** [comment]
Looking over the `constants.js` and MDN's `globalCompositeOperation` documentation, we believe that p5js's REPLACE should actually be the gCO `copy` and that p5js's BLEND should be `source-over`. Can someone confirm or deny this? We're not too familiar with what these blendModes _should_ be doing.

**Comment by tstefanich** [comment]
I need to find where it is in p5.js but these blendmodes that are in all caps just as REPLACE NORMAL ect are being translated into the proper web equivalents. These are ported over from how Processing did things. 

https://p5js.org/reference/#/p5/blendMode

ANd I believe the blendmode function actually supports both but I have not tested this. But I think thats right when I looked into this a few weeks ago.

**Comment by giltayar** [comment]
The place where this is in `constants.js`, where BLEND is mapped to `normal`, and REPLACE is mapped `source-over`. But according to [MDN](http://p5js.org/reference/#/p5/blendMode), and from what I understand from the [p5js reference on blendMode](https://p5js.org/reference/#/p5/blendMode), BLEND should be `source-over` and REPLACE should be `copy`, because according to the MDN docs, the difference is that `copy` ignores alpha, and `source-over` doesn't, and according to the p5 reference, REPLACE ignores alpha, and BLEND doesn't.

If my thinking is correct, we'll do the change and submit a PR.

**Comment by mlarghydracept** [comment]
Fixed with PR #1839 as per @giltayar 's suggestion.

**Comment by lmccart** [comment]
fixed with https://github.com/processing/p5.js/pull/1839
